File: backend-python/api/schemas/logs.py
# backend-python/api/schemas/logs.py
"""Schemas for log browsing and LLM call telemetry API endpoints."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class LogFileInfo(BaseModel):
    """Metadata for one log file in the logs directory."""

    name: str
    size: int
    updated_at: int


class LogListResponse(BaseModel):
    """Response wrapper for the log file listing endpoint."""

    files: list[LogFileInfo]


class LogFileContent(BaseModel):
    """Structured log file payload with parsed JSON or raw text."""

    name: str
    size: int
    updated_at: int
    is_json: bool
    content: Any | None = None
    raw: str | None = None


class LlmCallLogItem(BaseModel):
    """Summary row for one persisted LLM API call."""

    id: str
    provider: str
    api: str
    component: str
    status: str
    model_id: str | None = None
    job_id: str | None = None
    workflow_run_id: str | None = None
    task_run_id: str | None = None
    attempt: int | None = None
    latency_ms: int | None = None
    finish_reason: str | None = None
    input_tokens: int | None = None
    output_tokens: int | None = None
    total_tokens: int | None = None
    error_detail: str | None = None
    has_payload: bool = False
    created_at: int


class LlmCallLogListResponse(BaseModel):
    """Response wrapper for paged/sliced LLM call log summaries."""

    logs: list[LlmCallLogItem]


class LlmCallLogDetailResponse(BaseModel):
    """Detailed payload for one LLM call, including excerpts and snapshots."""

    log: LlmCallLogItem
    params_snapshot: dict[str, Any] | None = None
    request_excerpt: str | None = None
    response_excerpt: str | None = None
    payload_json: Any | None = None
    payload_raw: str | None = None

File: frontend/src/api/logs.ts
// src/api/logs.ts
import { API_BASE, apiFetch } from "./client";

export interface LlmCallLogItem {
    id: string;
    provider: string;
    api: string;
    component: string;
    status: "success" | "error" | string;
    model_id?: string | null;
    job_id?: string | null;
    workflow_run_id?: string | null;
    task_run_id?: string | null;
    attempt?: number | null;
    latency_ms?: number | null;
    finish_reason?: string | null;
    input_tokens?: number | null;
    output_tokens?: number | null;
    total_tokens?: number | null;
    error_detail?: string | null;
    has_payload: boolean;
    created_at: number;
}

export interface LlmCallLogListResponse {
    logs: LlmCallLogItem[];
}

export interface LlmCallLogDetailResponse {
    log: LlmCallLogItem;
    params_snapshot?: Record<string, unknown> | null;
    request_excerpt?: string | null;
    response_excerpt?: string | null;
    payload_json?: unknown;
    payload_raw?: string | null;
}

export async function fetchLlmCallLogs(options?: {
    limit?: number;
    component?: string;
    status?: "success" | "error";
}): Promise<LlmCallLogItem[]> {
    const params = new URLSearchParams();
    if (options?.limit) {
        params.set("limit", String(options.limit));
    }
    if (options?.component) {
        params.set("component", options.component);
    }
    if (options?.status) {
        params.set("status", options.status);
    }
    const qs = params.toString();
    const url = `${API_BASE}/api/logs/llm-calls${qs ? `?${qs}` : ""}`;
    const res = await apiFetch(url, {
        headers: { Accept: "application/json" },
    });
    const data = (await res.json()) as LlmCallLogListResponse;
    return data.logs ?? [];
}

export async function fetchLlmCallLog(
    id: string,
): Promise<LlmCallLogDetailResponse> {
    const res = await apiFetch(
        `${API_BASE}/api/logs/llm-calls/${encodeURIComponent(id)}`,
        {
            headers: { Accept: "application/json" },
        },
    );
    return res.json() as Promise<LlmCallLogDetailResponse>;
}

export async function deleteLlmCallLog(id: string): Promise<void> {
    await apiFetch(`${API_BASE}/api/logs/llm-calls/${encodeURIComponent(id)}`, {
        method: "DELETE",
        headers: { Accept: "application/json" },
    });
}

export async function clearLlmCallLogs(): Promise<number> {
    const res = await apiFetch(`${API_BASE}/api/logs/llm-calls`, {
        method: "DELETE",
        headers: { Accept: "application/json" },
    });
    const data = (await res.json()) as { deleted?: number };
    return data.deleted ?? 0;
}
